### PR TITLE
Fix GetKey for sets (plus some related changes)

### PR DESCRIPTION
**Pull request by shesek**
- The first commit is the simplest fix for a bug where the fingerprint wasn't compared correctly.

- The second & third commits are optional refactoring to reuse `Xpriv::get_key` for `$set<Xpriv>::get_key`, so the Xpriv matching logic only has to be maintained in one place.

- The forth commit adds support for signing with `Xpriv`s that are direct children of the `KeySource` -- possibly what the original (buggy) code author had in mind?

Of course, feel free to take just the first commit if the others seem unnecessary. The last one is kind of meh, not sure if really useful.

Note that multi-`Xpriv` signing does not actually work until #2850 is addressed too.

### CODE REVIEW

**Comment by apoelstra** [comment]
> But I personally find the entire API quite sketchy. The `KeySource` compares by 4-byte fingerprint only which seems small enough for me to be concerned about collisions,

Key fingerprints are pretty weird -- they are trivial to collide on purpose but they won't collide accidentally within any single wallet (when generating random keys you get 50% chance of collision after generating 65536 keys). So their purpose is for an individual wallet to be able to look up a key in its internal storage.

PSBT itself has a global map (and a per-input field which serves much the same purpose) mapping *full public keys* to fingerprints and paths. Since this map is keyed on full 256-bit keys (actually, full xpubs which are even bigger) it won't have collisions. But its purpose is to provide instructions to wallets "here is how to get this key", and the `KeySource` type is those instructions.

Now, a malicious actor can collide somebody else's fingerprints to put an invalid entry in the table, but from the wallet's point of view this is no different from any other invalid entry: the wallet takes the keysource, derives a key, finds that it doesn't match what's in the table, and assumes that this must be somebody else's key.

Summarizing -- we have a xpub -> keysource map in PSBT but for wallet applications we want a keysource -> xpub map (and a sanity check....which doesn't seem to exist here; it seems like when we find a matching keysource we just assume it's correct and return it).

To invert the map you need to do the iteration at least once. Though for some bizarre reason, rather than inverting the map we just store a set ... which has most of the memory cost of caching the inverted map but with none of the performance benefit. Very strange. I can see why you say "laughable".

Now, up to some renaming of types and modules, this code dates to #103 almost six years ago. The API is really weird and should be rewritten (which we are working on, outside this crate). But much of the weirdness is inherent to the weirdness of keysources. So I don't want to spend too much rust-bitcoin maintainer time trying to get this particular aspect of PSBT right.

**Comment by apoelstra** [comment]
TLDR I think we should accept this PR which fixes a bug, doesn't make anything worse, and makes the existing weird code at least look a little nicer.

**Comment by apoelstra** [comment]
>Though for some bizarre reason, rather than inverting the map we just store a set ...

Lol, re-reading https://github.com/rust-bitcoin/rust-bitcoin/pull/2850 I was reminded that in addition to the above weirdness, this code doesn't even work because we switched to using `BTreeSet`s but the keys aren't `Ord`, so it's impossible to use these. What a mess.

But I stand by my "let's just accept this marginal improvement" conclusion.

**Review by apoelstra — APPROVED** [review]
ACK b593c886e34e89b7c628e9b34de84a0e20e805af successfully ran local tests

**Comment by shesek** [comment]
> The API is really weird and should be rewritten (which we are working on, outside this crate)

Where is this happening? Related to PSBTv2?

I'm sure there are aspects I'm unaware of but fwiw the API hasn't been extremely weird for me. :) I wrote a bit more about my usage over at https://github.com/rust-bitcoin/rust-bitcoin/pull/2850#issuecomment-2350948311.

**Comment by tcharding** [comment]
> The API is really weird and should be rewritten (which we are working on, outside this crate)

Are you referring to https://crates.io/crates/psbt-v2 @apoelstra? If so I personally am no where near close to working on rewriting the signing API. Currently it has the code from `bitcion::psbt` imported from `v0.32.2` (I think I did `.2`).

I personally would be most happy if `psbt` in `rust-bitcoin` was deprecated and work was done over in `pstb-v2` crate. But if that doesn't happen then I don't have much motivation to work on that crate and no one else is _really_ committed because everyone is using `pstbt` from `bitcoin` and `miniscript`.

When `primitives` is done and we invert the dependency between `miniscript` and `bitcoin` we could consider importing it back in and remove `psbt` or just remove `psbt` and leave it separate.

**Review by tcharding — APPROVED** [review]
ACK b593c886e34e89b7c628e9b34de84a0e20e805af

**Comment by tcharding** [comment]
Thanks for the contribution. FWIW this was time consuming to review because it took me a while to work out

1. `GetKey` is not tested for sets (its actually broken like you point out elsewhere)
2. Logic in `GetKey` for sets of `Xpriv` does not match logic in `GetKey` for `Xpriv` (and wasn't caught because of (1)
3. PR adds additional functionality so that looked up keysource can have two different meanings 

The `GetKey for Xpriv` impl could probably do with some docs at some stage, won't hold this PR up for it though.

Thanks again!

**Comment by shesek** [comment]
> FWIW this was time consuming to review because

Could've probably done a better job communicating some of these points.

Thanks for merging!

**Comment by tcharding** [comment]
No stress man, appreciate you PR'ing!

**Comment by tcharding** [comment]
Hey @shesek in case you are interested we've recently pulled the `psbt` module out of `rust-bitcoin` and release `rust-psbt-v0` with the hope of iterating on PSBT stuff faster. I'm backporting your bug fixes just now :)

https://crates.io/crates/psbt-v0

**Pull request merged**
